// vite-plugin-monaco-editor-esm/dist/workerMiddleware.js
import { getWorks, isCDN, resolveMonacoPath } from './index.js';
import * as esbuild from 'esbuild';
import * as fs from 'fs';
import * as path from 'path';
export function getFilenameByEntry(entry) {
    entry = path.basename(entry, 'js');
    return entry + '.bundle.js';
}
export const cacheDir = 'node_modules/.monaco/';
export function getWorkPath(works, options, config) {
    const workerPaths = {};
    for (const work of works) {
        if (isCDN(options.publicPath ?? '')) {
            workerPaths[work.label] = options.publicPath + '/' + getFilenameByEntry(work.entry);
        }
        else {
            workerPaths[work.label] =
                config.base + options.publicPath + '/' + getFilenameByEntry(work.entry);
        }
    }
    if (workerPaths['typescript']) {
        // javascript shares the same worker
        workerPaths['javascript'] = workerPaths['typescript'];
    }
    if (workerPaths['css']) {
        // scss and less share the same worker
        workerPaths['less'] = workerPaths['css'];
        workerPaths['scss'] = workerPaths['css'];
    }
    if (workerPaths['html']) {
        // handlebars, razor and html share the same worker
        workerPaths['handlebars'] = workerPaths['html'];
        workerPaths['razor'] = workerPaths['html'];
    }
    return workerPaths;
}
export function workerMiddleware(middlewares, config, options) {
    const works = getWorks(options);
    if (!works) {
        throw new Error('No work definition found.');
    }
    // clear cacheDir
    if (fs.existsSync(cacheDir)) {
        fs.rmSync(cacheDir, { recursive: true, force: true });
    }
    for (const work of works) {
        middlewares.use(config.base + options.publicPath + '/' + getFilenameByEntry(work.entry), function (req, res, next) {
            if (!fs.existsSync(cacheDir + getFilenameByEntry(work.entry))) {
                esbuild.buildSync({
                    entryPoints: [resolveMonacoPath(work.entry)],
                    bundle: true,
                    outfile: cacheDir + getFilenameByEntry(work.entry),
                });
            }
            const contentBuffer = fs.readFileSync(cacheDir + getFilenameByEntry(work.entry));
            res.setHeader('Content-Type', 'text/javascript');
            res.end(contentBuffer);
        });
    }
}
